libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1719,3 +1740,189 @@ Type const& IRGeneratorForStatements::type(Expression const& _expression)
 	solAssert(_expression.annotation().type, "Type of expression not set.");
 	return *_expression.annotation().type;
 }
+
+bool IRGeneratorForStatements::visit(TryStatement const& _tryStatement)
+{
+	solAssert(!m_returnInfo.has_value(), "");
+
+	m_returnInfo.reset();
+	_tryStatement.externalCall().accept(*this);
+	solAssert(m_returnInfo.has_value(), "");
+
+	auto const trySuccessCondition = m_context.trySuccessConditionVariable(_tryStatement.externalCall());
+
+	m_code << "switch iszero(" << trySuccessCondition << ")\n";
+
+	m_code << "case 0 { // success case\n";
+	TryCatchClause const& successClause = *_tryStatement.clauses().front();
+	decodeTryCallReturnParameters(successClause);
+	m_returnInfo.reset();
+	successClause.block().accept(*this);
+	m_code << "}\n";
+
+	m_code << "default { // failure case\n";
+	handleCatch(_tryStatement);
+	m_code << "}\n";
+
+	return false;
+}
+
+string IRGeneratorForStatements::decodeReturnParameters()
+{
+	ABIFunctions abi(m_context.evmVersion(), m_context.revertStrings(), m_context.functionCollector());
+
+	return Whiskers(R"(
+		<?dynamicReturnSize>
+			// copy dynamic return data out
+			returndatacopy(<pos>, 0, returndatasize())
+		</dynamicReturnSize>
+
+		// update freeMemoryPointer according to dynamic return size
+		mstore(<freeMemoryPointer>, add(<pos>, and(add(<returnSize>, 0x1f), not(0x1f))))
+
+		// decode return parameters from external try-call into <retVars>
+		let <retVars> := <abiDecode>(<pos>, add(<pos>, <returnSize>))
+	)")
+	("retVars", IRVariable(m_returnInfo->functionCall).commaSeparatedList())
+	("pos", m_returnInfo->returndataVariable)
```

Comment:
This really has to go into a yul function parameter.

Reply:
This is returning yul code, not a function. So far (IMHO) it still makes sense the way it is, unless I am moving the code into a yul function (and then move the function into YulUtilFunctions API). I remember I tried that. I can see check again I failed last time.

Reply:
@chriseth: The calling AST node (FunctionCall) is already (optionally) stored in `ReturnInfo`; I've added a solAssert() underneath to ensure it contains the expected one.

Reply:
I'm not sure it's worth putting this into YulUtilFunctions. At least I'd like to get some feedback from @ekpyron wrt. best practice next week (seperate PR then?).

Reply:
It is _much_ cleaner if this is a proper function.

Reply:
The code here should be `let <retVars> := <decodeReturnParameters>(<pos>)`

Reply:
This is also what @ekpyron suggested 14 days ago: https://github.com/ethereum/solidity/pull/8246/files#r390886194

Reply:
@ekpyron once said recommendations aren't "mandatory", but I got his point and tried it back in time. I've now found a way to implement it (back in time i failed) :-)